Conditionally queue a task to notify

domenic opened the pull request:
<del>This is option 5 in @bzbarsky's https://github.com/domenic/unhandled-rejections-browser-spec/issues/2#issuecomment-120582936. Some test cases change since now we allow a bit more delay before attachment.</del><ins>Option 3 in play now, plus a few other changes.</ins>

See more discussion over in that thread. If we decide it is a good change, then review of this would be much appreciated before pulling it in, especially in the form of @jeisinger trying it in Blink :)

domenic committed:
Conditionally queue a task to notify

This is option 5 in @bzbarsky's https://github.com/domenic/unhandled-rejections-browser-spec/issues/2#issuecomment-120582936. Some test cases change since now we allow a bit more delay before attachment.

domenic committed:
Add tests for subtle task interactions

See https://github.com/domenic/unhandled-rejections-browser-spec/issues/2#issuecomment-121121695 and also https://github.com/domenic/unhandled-rejections-browser-spec/issues/2#issuecomment-121301595

domenic committed:
Fix misplaced Done label

See https://github.com/domenic/unhandled-rejections-browser-spec/issues/2#issuecomment-121134345

domenic committed:
Switch to a snapshot strategy

This is 3 from @bzbarsky's https://github.com/domenic/unhandled-rejections-browser-spec/issues/2#issuecomment-120582936. See subsequent comments to find out why we converged on that.

domenic committed:
Integrate notes into main text

Switch to "not handled" instead of "unhandled" for the definition of what signifier is used for the console, in order to better match HTML's existing term for exceptions.

domenic commented:
@bzbarsky @jeisinger I think this is ready for review. Updated to option 3. You can view the whole thing at https://github.com/domenic/unhandled-rejections-browser-spec/blob/more-delay/README.md#changes-to-html

domenic committed:
Fire rejectionhandled from within a task

jeisinger commented on tests/resources/promise-rejection-events.js:
```diff
@@ -464,47 +464,87 @@ async_test(function(t) {
     p.then(unreached, function() {});
   });
-}, 'delayed handling: a task delay before attaching a handler does not prevent unhandledrejection');
+}, 'delayed handling: a task delay before attaching a handler prevents unhandledrejection');
```
why? The microtask checkpoint already scheduled a task to report the event. When the handler is added, the rejected promise is no longer in the list of about to be notified promises, so we'll just queue a task to notify about the addition of the handler as well.

jeisinger commented on tests/resources/promise-rejection-events.js:
```diff
@@ -586,6 +673,121 @@ async_test(function(t) {
+  });
+
+  function l(ev) {
+    sequenceOfEvents.push(ev.promise);
```
we need to protect this by if (ev.promise === p1 || ev.promise === p2) {} here and in the other new handlers you added

domenic replied:
I keep getting caught out by how testharness.js allows multiple async tests in parallel. So evil.

jeisinger commented on tests/resources/promise-rejection-events.js:
```diff
@@ -586,6 +673,121 @@ async_test(function(t) {
+
+  var p1 = Promise.reject();
+  var p2;
+  postMessageTask(function() {
```
on workers, the MessageChannel port postMessage thing bounces to the main thread, and as such there are no ordering guarantees wrt to the promise events

jeisinger commented on tests/resources/promise-rejection-events.js:
```diff
@@ -446,14 +446,14 @@ async_test(function(t) {
   var p;
 
-  onUnhandledSucceed(t, e, function() { return p; });
+  onUnhandledFail(t, function() { return p; });
```
for some odd reason, both the old and the new version of this test pass... not sure what it tests

jeisinger replied:
so this test fails. I think the test is wrong... 

the task to send the unonunhandled event is created in the microtask checkpoint. the task that attaches a handler then runs before, but at that point, the task for the onunhandled event already exists, and it won't be cancelled.

domenic replied:
Right, the task already exists, but the list may have been modified since the task was posted: "If the about-to-be-notified rejected promises list contains promise, remove promise from the about-to-be-notified rejected promises list and return."

jeisinger replied:
at the point where the task is posted, the about-to-be-notified list is
copied and wiped.

On Fri, Oct 2, 2015 at 1:57 PM Domenic Denicola notifications@github.com
wrote:

> In tests/resources/promise-rejection-events.js
> https://github.com/domenic/unhandled-rejections-browser-spec/pull/13#discussion_r41013947
> :
> 
> >  //
> > 
> >  async_test(function(t) {
> >    var e = new Error();
> >    var p;
> > -  onUnhandledSucceed(t, e, function() { return p; });
> > -  onUnhandledFail(t, function() { return p; });
> 
> Right, the task already exists, but the list may have been modified since
> the task was posted: "If the about-to-be-notified rejected promises list
> contains promise, remove promise from the about-to-be-notified rejected
> promises list and return."
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/domenic/unhandled-rejections-browser-spec/pull/13/files#r41013947
> .

domenic replied:
Oh, hmm, the spec definitely makes a copy as-is. But that seems counter to what we agreed, and what the OP of this pull request says ("option 3"). Argh, trying to reload our context...

Does it work to just use the same global list throughout all algorithms, and clear it after the task loop? Or does that have possible reentrancy issues? I guess we could do what was suggested in option 3 directly, and re-check each promise's internal state to see if it's been handled.

jeisinger replied:
hum, if we don't copy the list, it feels really random when the events are
sent and when they aren't.

On Fri, Oct 2, 2015 at 2:04 PM Domenic Denicola notifications@github.com
wrote:

> In tests/resources/promise-rejection-events.js
> https://github.com/domenic/unhandled-rejections-browser-spec/pull/13#discussion_r41014524
> :
> 
> >  //
> > 
> >  async_test(function(t) {
> >    var e = new Error();
> >    var p;
> > -  onUnhandledSucceed(t, e, function() { return p; });
> > -  onUnhandledFail(t, function() { return p; });
> 
> Oh, hmm, the spec definitely makes a copy as-is. But that seems counter to
> what we agreed, and what the OP of this pull request says ("option 3").
> Argh, trying to reload our context...
> 
> Does it work to just use the same global list throughout all algorithms,
> and clear it after the task loop? Or does that have possible reentrancy
> issues? I guess we could do what was suggested in option 3 directly, and
> re-check each promise's internal state to see if it's been handled.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/domenic/unhandled-rejections-browser-spec/pull/13/files#r41014524
> .

domenic replied:
Why? I thought that was the entire content of the discussion in #2. We should not send events if the promise has been handled by the time we _would_ send the events.

jeisinger replied:
actually, you're right.

the tests then pass on document but not on worker, since the
postMessageTask() thing

I've updated the CL with tests + implementation

if you update your spec text, and import my tests, we should be good to go

On Fri, Oct 2, 2015 at 2:10 PM Domenic Denicola notifications@github.com
wrote:

> In tests/resources/promise-rejection-events.js
> https://github.com/domenic/unhandled-rejections-browser-spec/pull/13#discussion_r41014865
> :
> 
> >  //
> > 
> >  async_test(function(t) {
> >    var e = new Error();
> >    var p;
> > -  onUnhandledSucceed(t, e, function() { return p; });
> > -  onUnhandledFail(t, function() { return p; });
> 
> Why? I thought that was the entire content of the discussion in #2
> https://github.com/domenic/unhandled-rejections-browser-spec/issues/2.
> We should not send events if the promise has been handled by the time we
> _would_ send the events.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/domenic/unhandled-rejections-browser-spec/pull/13/files#r41014865
> .

domenic committed:
Fix onUnhandledFail to actually work

As pointed out by @jeisinger in https://github.com/domenic/unhandled-rejections-browser-spec/pull/13#discussion_r40895565 previously onUnhandledFail did not actually cause failures. Eek!

jeisinger commented on tests/resources/promise-rejection-events.js:
```diff
@@ -642,12 +844,12 @@ function onUnhandledSucceed(t, expectedReason, expectedPromiseGetter) {
   var unhandled = function(evt) {
     if (evt.promise === expectedPromiseGetter()) {
-      t.unreached_func('unhandledrejection event is not supposed to be triggered');
+      t.assert_unreached('unhandledrejection event is not supposed to be triggered');
```
actually, I now get this error:

Harness Error. harness_status.status = 1 , harness_status.message = Uncaught TypeError: t.assert_unreached is not a function

domenic replied:
Argh, I thought I tested this. The following should work in theory:

``` js
t.step(function() {
  assert_unreached('...');
});
t.done(); // maybe this line is optional
```

if not I can do better when I get in to work in a few hours; just woke up.

jeisinger replied:
cool, that works.

however, now the tests pass no longer...

jeisinger commented on tests/resources/promise-rejection-events.js:
```diff
@@ -464,47 +464,87 @@ async_test(function(t) {
+      p.catch(function() {});
+    });
+  });
+}, 'delayed handling: postMessageTask after promise creation/rejection, plus promise microtasks, is not too late to ' +
```
same here, it's too late

jeisinger commented on tests/resources/promise-rejection-events.js:
```diff
@@ -464,47 +464,87 @@ async_test(function(t) {
+    });
+  });
+  p = Promise.reject(e);
+}, 'delayed handling: postMessageTask before promise creation/rejection, plus many promise microtasks, is not too ' +
```
and here

jeisinger commented on tests/resources/promise-rejection-events.js:
```diff
@@ -464,47 +464,87 @@ async_test(function(t) {
-  ensureCleanup(t, unhandled, handled);
+    });
+  });
+}, 'delayed handling: postMessageTask after promise creation/rejection, plus many promise microtasks, is not too ' +
```
... and here

jeisinger committed:
Fixes from running the tests in Blink

- Guard better against concurrent execution of async tests
- Don't run tests that rely on postMessageTask inside workers

domenic committed:
Update the spec to not copy the list

To actually implement the conclusion of #2, this is more correct.
